File: Napakalaki_Ruby/lib/player.rb
#encoding: utf-8
#  To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

  require_relative 'treasure.rb'
  require_relative 'dice.rb'
  
module Napakalaki
  
  class Player
    
    attr_reader :dead, :name, :level, :pending_bad_consequence, :hidden_treasures, :visible_treasures, :is_cultist
    
    def initialize(name)
      @name = name
      @dead = true
      @level = 1;
      @visible_treasures = Array.new
      @hidden_treasures = Array.new
      @is_cultist = false
    end
    
    def is_dead()
      return @dead
    end    
    
    def combat(monster)
      my_level = get_combat_level()
      if is_cultist()
        mlevel = monster.get_special_value
      else
        mlevel = monster.get_basic_value      
      end
      if my_level > mlevel then
        apply_prize(monster)
        if @level >= 10 then
          result = CombatResult::WinAndWinGame
        else
          result = CombatResult::Win
        end
      else
        dice = Dice.instance
        escape = dice.next_number()
        if escape < 5 then
          dead = monster.kills()
          if dead
            die()
            result = CombatResult::LoseAndDie
          else
            mal_r = monster.mal_rollo()            
            apply_bad_consequence(mal_r)
              if(shouldConvert())
                result = CombatResult::LoseAndConvert;
              else
                result = CombatResult::Lose   
              end
          end
          else
          result = CombatResult::LoseAndEscape
        end
      end
      discard_necklace_visible()
      return result
    end
    
    def make_treasure_visible(t)
      can_i = can_make_treasure_visible(t)
      if can_i then
        @visible_treasures << t
        @hidden_treasures.delete(t)
      end
    end
    
    def discard_visible_treasure(t)
      @visible_treasures.delete(t)
      if @pending_bad_consequence != nil && !@pending_bad_consequence.is_empty()
        @pending_bad_consequence.substract_visible_treasures(t)
        if(@pending_bad_consequence.n_visible_treasures() > 0)
          @pending_bad_consequence.n_visible_treasures(@pending_bad_consequence.n_visible_treasures()-1)
        end
      end
      die_if_no_treasures()
    end
    
    def discard_hidden_treasure(t)
      @hidden_treasures.delete(t)
      if @pending_bad_consequence != nil && !@pending_bad_consequence.is_empty() then
        @pending_bad_consequence.substract_hidden_treasures(t)
        if(@pending_bad_consequence.n_hidden_treasures() > 0)
          @pending_bad_consequence.n_hidden_treasures(@pending_bad_consequence.n_hidden_treasures()-1)
        end
      end
      die_if_no_treasures()
    end
    
    def buy_levels(visible, hidden)
      levels_i_can_buy = 0;
      levels_i_can_buy += compute_gold_coins_value(visible) 
      levels_i_can_buy += compute_gold_coins_value(hidden) 
      can_i = can_i_buy_levels(levels_i_can_buy)
      increment_levels(levels_i_can_buy) if can_i
      @visible_treasures.delete(visible) 
      @hidden_treasures.delete(hidden) 
      dealer = CardDealer.instance 
      visible.each() do |t|
        dealer.give_treasure_back(t) 
      end
      hidden.each() do |t|
        dealer.give_treasure_back(t) 
      end
      return can_i
    end
    
    def valid_state()
      v_state = false
      if @pending_bad_consequence == nil && @hidden_treasures.length() <= 4
      v_state = true
      else
        v_state = true if @pending_bad_consequence.is_empty() && @hidden_treasures.length() <= 4
      end
    end
    
    def init_treasures
      dealer = CardDealer.instance
      dice = Dice.instance
      bring_to_life()
      treasure = dealer.next_treasure
      @hidden_treasures << treasure
      number = dice.next_number
      if number == 6
        treasure = dealer.next_treasure
        @hidden_treasures << treasure
        treasure = dealer.next_treasure
        @hidden_treasures << treasure 
      else
        if number > 1
        treasure = dealer.next_treasure 
        @hidden_treasures << treasure 
        end
      end
    end
          
    def has_visible_treasures()
      return @visible_treasures.size() > 0
    end
    
    def to_s
    " #{@name}. Nivel: #{@level}. Nivel de Combate: #{get_combat_level()}. Muerto: #{@dead}."
    end
      
    private
    
    def bring_to_life()
      @dead = 0
    end
    
    def get_combat_level()
      l = 0
      collar = false
      @visible_treasures.each do |t|
        if t.type == TreasureKind::Necklace then
          collar = true
          break
        end
      end
      
      @visible_treasures.each do|t|
        if( collar == true )
          l += t.max_bonus
        else
          l += t.min_bonus  
        end        
      end
    return @level + l
    end
    
    def set_pending_bad_consequence(b)
      @pending_bad_consequence = b
    end
    
    def increment_levels(l)
      @level = @level + l
    end
    
    def decrement_levels(l)
      @level = @level - l
      @level = 1 if @level < 1 
    end
    
    def die_if_no_treasures()
      if( @hidden_treasures.size() == 0 && @visible_treasures.size() == 0 )
        @dead = 1
      end
    end
    
    def discard_necklace_visible()
      @visible_treasures.each do |t|       
        if t.type() == TreasureKind::Necklace
          CardDealer.give_treasure_back(t)
          @visible_treasures.delete_at(t) 
        end
      end
    end
  
    def die()
      @level = 1
      dealer = CardDealer.instance
      @visible_treasures.each do |t|
        dealer.give_treasure_back(t)
      end
      @visible_treasures.clear()
      @visible_treasures.each do |t| 
        dealer.give_treasure_back(t)
      end
      @hidden_treasures.clear
      die_if_no_treasures()
    end
    
    def compute_gold_coins_value(t)
      coins = 0
      t.each() do |k|
        coins = coins + k.gold_coins()
      end
      levels = coins / 1000
      return levels
    end    
   
    def can_i_buy_levels(l)        
      can_i = false
      can_i = true if @level + l < 10
      return can_i
    end
    
    def apply_prize(monster)
      n_levels = monster.buen_rollo().level()
      increment_levels(n_levels)
      n_treasures = monster.get_treasures_gained()
      if n_treasures > 0
        dealer = CardDealer.instance
        while n_treasures > 0
          t = dealer.next_treasure
          @hidden_treasures << t
          n_treasures = n_treasures - 1 
        end
      end
    end
    
    def apply_bad_consequence(bad_consequence)
      n_levels = bad_consequence.levels()      
      decrement_levels(n_levels)
      @pending_bad_consequence = bad_consequence.adjust_to_fit_treasure_list(@visible_treasures, @hidden_treasures) 
      set_pending_bad_consequence(@pending_bad_consequence)
    end
    
    def can_make_treasure_visible(t)
      result = true
          if (@visible_treasures.size() < 4) 
            type = t.type();
                if type == TreasureKind::OneHand 
                  vt = @visible_treasures.clone
                  num = 0
                  vt.each do |k| 
                    num = num + 1 if (k.type() == TreasureKind::OneHand)
                  end
                  result = false if (num == 2)
                  @visible_treasures.each do |k| 
                    result = false if k.type() == TreasureKind::BothHand                    
                  end
                else
                  num = 0
                  @visible_treasures.each do |k| 
                    result = false if (k.type() == t.type())
                  end
                end
          else
            result = false
          end
      return result
    end
     
    def how_many_visible_treasures(t_kind)
      cont = 0      
      @visible_treasures.each() do |i|
        t_kind.each() do |j|
          cont = cont + 1 if ( i == j )
        end
      end
      return cont
    end  
        
    def should_convert()
        convert = false
        dice = Dice.instance
        if dice.next_number == 6
            convert = true;
        end
        return convert;
    end
  end
end

File: Napakalaki/src/napakalaki/Player.java
/**
 *
 */

package napakalaki;

import static java.lang.Math.min;
import java.util.ArrayList;

public class Player {
    private boolean dead = true;
    private int level=-1;
    private String name;
    private ArrayList<Treasure> HiddenTreasures;
    private ArrayList<Treasure> VisibleTreasures;
    private BadConsequence pendingBadConsequence;
    protected boolean cultist = false;
    
    @Override
    public String toString(){
        return name;
    }
    
    public Player(Player player){
        name = player.getName();
        level = player.getLevel();
        dead = player.isDead();
        HiddenTreasures = (ArrayList<Treasure>) player.getHiddenTreasures().clone();
        VisibleTreasures = (ArrayList<Treasure>) player.getVisibleTreasures().clone();
        pendingBadConsequence = player.getPendingBadConsequence();
    }
    
    public Player(String name){
        this.name = name;
        level = 1;
        dead = false;
        HiddenTreasures = new ArrayList<>();
        VisibleTreasures = new ArrayList<>();
        pendingBadConsequence = new BadConsequence("Primer Turno",0,0,0);
    }
    
    /*Devuelve la vida al jugador, modificando el atributo correspondiente.*/
    private void bringToLife(){
        dead = false;
    } 
    /*Incrementa el nivel del jugador en l niveles*/
    private void incrementLevels(int l){
        level = level + l;
    }
    /*Decrementa el nivel del jugador en l niveles*/
    private void decrementLevels(int l){
        level = level - l;
        if(level < 1) level = 1;
    }
   /*Asigna el mal rollo al jugador, dándole valor a su atributo 
    pendingBadConsequence.*/ 
    private void setPendingBadConsequence(BadConsequence bad) {
        pendingBadConsequence = bad;
    }
    //Mata al jugador
    private void die(){
        CardDealer cd = CardDealer.getInstance();
        for(int i = 0; i < VisibleTreasures.size(); i++){
            Treasure treasure = cd.nextTreasure();
            cd.giveTreasureBack(treasure);
        }
        VisibleTreasures.clear();
        for(int i = 0; i < HiddenTreasures.size(); i++){
            Treasure treasure = cd.nextTreasure();
            cd.giveTreasureBack(treasure);
        }
        HiddenTreasures.clear();
        level = 1;
        dead = true;
    }
    
    /*Si el jugador tiene equipado el tesoro tipo NECKLACE, se lo pasa al 
    CardDealer y lo elimina de sus tesoros visibles.*/
    private void discardNecklaceIfVisible(){
        CardDealer cd = CardDealer.getInstance();
        boolean have_necklace = false;
        int N = VisibleTreasures.size();
        
        for(int i = 0; i < N ; i++){
            if(VisibleTreasures.get(i).getType().equals(TreasureKind.NECKLACE)){
                cd.giveTreasureBack(VisibleTreasures.get(i));
                VisibleTreasures.remove(i);
            }
        }
    }
    
    /*Cambia el estado de jugador a muerto si no tiene ni tesoros visibles ni 
    ocultos, modificando el correspondiente atributo.*/
    private void dieIfNoTreasures(){
        if(VisibleTreasures.isEmpty() && HiddenTreasures.isEmpty())
            die();
    }
    
    /*Calcula y devuelve los niveles que puede comprar el jugador con la lista 
    t de tesoros.*/
    private int computeGoldCoinsValue(ArrayList<Treasure> t){
        int maxcoins = 0, N=t.size(),canbuylevels;
        for(int i = 0; i < N; i++){
            maxcoins += t.get(i).getGoldCoins();
        }
        canbuylevels = maxcoins/1000; //Division entera
        return canbuylevels;
    }
    
    /*Devuelve true si con los niveles que compra no gana la partida y 
    false en caso contrario.*/
    private boolean canIBuyLevels(int l){
        Napakalaki game = Napakalaki.getInstance();
        boolean stat;
        if(l > game.getCurrentMonster().getLevel()){
            stat = true;
        }
        else
            stat = false;
        return stat;
    }
    
    public void applyPrize(Prize p){
        int nlevels = p.getLevels();
        int nprize = p.getTreasures();
        Treasure t = null;
        incrementLevels(nlevels);
        for(int i = 0; i < min(nprize,4-HiddenTreasures.size()); i++){
            CardDealer cd = CardDealer.getInstance();
            t = cd.nextTreasure();
            HiddenTreasures.add(t);
        }
    }
    
/*El método combat() de la clase Player debe modificarse para que una vez que el
jugador activo haya combatido y en caso de que no haya podido huir pero siga 
vivo:
• Se envíe a sí mismo el mensaje shouldConvert()
• Si el mensaje devuelve true devuelva el valor LoseAndConvert*/
    
/*En el método combat() de Player se modificará para que, en vez de preguntar el
    nivel del monstruo al que se enfrenta al propio monstruo con el método 
    getLevel(), se preguntará a sí mismo con el método getOponentLevel(). El 
    método getOponentLevel() debe definirse en Player y redefinirse en 
    CultistPlayer para hacer uso de getBasicValue() y getSpecialValue() de la 
    clase Monster devolviendo respectivamente el nivel del monstruo al que se 
    enfrenta un jugador y un jugador sectario.*/
    
    public CombatResult combat(Monster m){
        CombatResult stat = null; 
        int myLevel = getCombatLevel();
        //int levelMonster = m.getLevel();
        int levelMonster = getOponentLevel();
        int dice_result;
        boolean amIdead;
        Dice dc = Dice.getInstance();
        if(myLevel>levelMonster){
            Prize prize = m.getPrize();
            applyPrize(prize);
            if(level<10){
                stat = CombatResult.WIN;
            }
            else
                stat = CombatResult.WINANDWINGAME;
        }
        else{
            dice_result = dc.nextNumber();
            boolean not_escape = dice_result<5;
            if(not_escape){
                BadConsequence bad = m.getBad();
                amIdead = bad.kills();
                if(amIdead==true){
                    die();
                    stat = CombatResult.LOSEANDDIE;
                }
                else{
                    applyBadConsequence(bad);
                    if(shouldConvert())
                        stat = CombatResult.LOSEANDCONVERT;                        
                    else
                        stat = CombatResult.LOSE;     
                }
            }
            else{
                stat=CombatResult.LOSEANDESCAPE;
            }
        }
        discardNecklaceIfVisible();
        return stat;
    }
    
    public void applyBadConsequence(BadConsequence bad){
        int nlevels = bad.getLevels();
        decrementLevels(nlevels);
        BadConsequence pendingBad = bad.adjustToFitTreasureLists(VisibleTreasures, 
                HiddenTreasures);
        setPendingBadConsequence(pendingBad);
    }
    
    public void makeTreasureVisible(Treasure t){
            getHiddenTreasures().remove(t);
            getVisibleTreasures().add(t);
    }
    
    /*Comprueba si el tesoro (t) se puede pasar de oculto a visible, según las 
    reglas del juego*/
    
    public boolean canMakeTreasureVisible(Treasure t){
        TreasureKind type = t.getType();
        boolean canMakeVisible=true; //Variable: -> Se puede hacer visible? 
        boolean isHandKind=false;
        int handCounter=0;
        boolean isEquipped=false;
        
        if(type.equals(TreasureKind.BOTHHANDS)||type.equals(TreasureKind.ONEHAND)){
            isHandKind = true;
        }
        
        if(isHandKind){
            for(int i=0; i<getVisibleTreasures().size() ; i++){
                if(getVisibleTreasures().get(i).getType().equals(TreasureKind.ONEHAND))
                    handCounter++;
                if(getVisibleTreasures().get(i).getType().equals(TreasureKind.BOTHHANDS))
                    handCounter += 2;
            }
            if(type.equals(TreasureKind.BOTHHANDS )){
                if(handCounter>0){
                    canMakeVisible=false;
                }
            }
            if(type.equals(TreasureKind.ONEHAND)){
               if(handCounter>1){
                    canMakeVisible=false;
                }
            }
        }
        else{
            for(int i=0; i<getVisibleTreasures().size(); i++){
                if(getVisibleTreasures().get(i).getType().equals(type)){
                    isEquipped=true;
                }
            }
        }
        return (canMakeVisible && !isEquipped);
    } 
    
    public void discardVisibleTreasure(Treasure t){
        CardDealer cd = CardDealer.getInstance();
        VisibleTreasures.remove(t);
        if((!pendingBadConsequence.isEmpty())){
            pendingBadConsequence.substractVisibleTreasure(t);
        }
        cd.giveTreasureBack(t);
        dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t){
        CardDealer cd = CardDealer.getInstance();
        HiddenTreasures.remove(t);
        if((!pendingBadConsequence.isEmpty())){
            pendingBadConsequence.substractHiddenTreasure(t);
        }
        cd.giveTreasureBack(t);
        dieIfNoTreasures();
    }
    
    public boolean buyLevels(ArrayList<Treasure> visible,
            ArrayList<Treasure> hidden){
        int levels = computeGoldCoinsValue(visible);
        levels += computeGoldCoinsValue(hidden);
        boolean canI = canIBuyLevels(levels+getCombatLevel());
        if (canI){
            incrementLevels(levels);
            for(int i = 0; i < visible.size(); i++){
                discardVisibleTreasure(visible.get(i));
            }
            for(int i = 0; i < hidden.size(); i++){
                discardHiddenTreasure(hidden.get(i));
            }
        }
        return canI;
    }

    /*Devuelve el nivel de combate del jugador, que viene dado por su nivel más 
    los bonus que le proporcionan los tesoros que tenga equipados, según las 
    reglas del juego.*/
    
    public int getCombatLevel() {
        int totallevel = level;
        boolean haveNecklace = false;
        
        for(int i = 0; i < VisibleTreasures.size();i++){
            if(TreasureKind.NECKLACE.equals(VisibleTreasures.get(i).getType())){
                haveNecklace = true;
            }
        }
        
        for(int i = 0; i < VisibleTreasures.size();i++){
            if(haveNecklace)
                totallevel += VisibleTreasures.get(i).getMaxBonus();
            else{
                totallevel += VisibleTreasures.get(i).getMinBonus();
            }
        }
        return totallevel;
    }
    
    /*Devuelve true cuando el jugador no tiene ningún mal rollo que cumplir
(pendingBadConsequence.isEmpty() == true) y false en caso contrario.*/
    
    public boolean validState(){
        boolean isValid = true;
        if(!pendingBadConsequence.isEmpty()){
            isValid = false;
        }
        if(HiddenTreasures.size()>4){
            isValid = false;
        }        
        return isValid;
    }
    
    public void initTreasures(){
        CardDealer cd = CardDealer.getInstance();        
        Dice dice = Dice.getInstance();
        Treasure t;
        if(isDead()){
            bringToLife();
        }
        int number = dice.nextNumber();
        if(number == 1){
            t = cd.nextTreasure();
            HiddenTreasures.add(t);
        }
        else if(number < 6){
            for(int i = 1; i <= 2;i++){
                t = cd.nextTreasure();
                HiddenTreasures.add(t);
            }
        }
        else if(number == 6){
             for(int i = 1; i <= 3;i++){
                t = cd.nextTreasure();
                HiddenTreasures.add(t);
            }
        }
    }
    
    /*Devuelve true si el jugador está muerto, false en caso contrario.*/
    
    public boolean isDead() {
        return dead;
    }
    
    public boolean isCultist(){
        return cultist;
    }
    /*Devuelve true cuando el jugador tiene algún tesoro visible y false en 
    caso contrario.*/
    
    public boolean hasVisibleTreasures(){
        boolean hastreasure = false;
        if(VisibleTreasures.size() > 0)
            hastreasure = true;
        return hastreasure;   
    }

    public ArrayList<Treasure> getVisibleTreasures() {
        return VisibleTreasures;
    }
    
    public ArrayList<Treasure> getHiddenTreasures() {
        return HiddenTreasures;
    }
    
    public String getName(){
        return name;
    }
    
    public int getLevel(){
        return level;
    }
    
    public BadConsequence getPendingBadConsequence(){
        return pendingBadConsequence;
    }
    
    public int getOponentLevel(){
        int mlevel;
        if (!isCultist())
            mlevel = Napakalaki.getInstance().getCurrentMonster().getBasicValue();
        else
            mlevel = Napakalaki.getInstance().getCurrentMonster().getSpecialValue();
        return mlevel;
    }
    
    /*En el método shouldConvert() de la clase Player se lanzará el dado y 
    devolverá true si se obtiene un 6 y false en caso contrario. Su redefinición
    en CultistPlayer devolverá siempre false.*/
    
    public boolean shouldConvert(){
        boolean convert = false;
        if(Dice.getInstance().nextNumber() == 6)
            convert = true;
        return convert;
    }
}